src/engine/shanghai.md

```diff
@@ -134,4 +148,64 @@ This method follows the same specification as [`engine_forkchoiceUpdatedV1`](./p
 
-  1. Client software **SHOULD** use the sum of the block's priority fees or any other algorithm to determine `blockValue`.
\ No newline at end of file
+  1. Client software **SHOULD** use the sum of the block's priority fees or any other algorithm to determine `blockValue`.
+
+### engine_getPayloadBodiesByHashV1
+
+#### Request
+
+* method: `engine_getPayloadBodiesByHashV1`
+* params:
+  1. `Array of DATA`, 32 Bytes - Array of `block_hash` field values of the `ExecutionPayload` structure
+* timeout: 10s
+
+#### Response
+
+* result: `Array of ExecutionPayloadBodyV1` - Array of [`ExecutionPayloadBodyV1`](#ExecutionPayloadBodyV1) objects.
+* error: code and message set in case an exception happens while processing the method call.
+
+#### Specification
+
+1. Given array of block hashes client software **MUST** respond with array of `ExecutionPayloadBodyV1` objects with the corresponding hashes respecting the order of block hashes in the input array.
+
+1. Client software **MUST** place responses in the order given in the request, using `null` for any missing blocks. For instance, if the request is `[A.block_hash, B.block_hash, C.block_hash]` and client software has data of payloads `A` and `C`, but doesn't have data of `B`, the response **MUST** be `[A.body, null, C.body]`.
+
+1. Client software **MAY NOT** respond to requests for finalized blocks by hash.
+
+1. Client software **MUST** set `withdrawals` field to `null` for bodies of pre-Shanghai blocks.
+
+1. This request maps to [`BeaconBlocksByRoot`](https://github.com/ethereum/consensus-specs/blob/dev/specs/phase0/p2p-interface.md#beaconblocksbyroot) in the consensus layer `p2p` specification. Callers must be careful to use the execution block hash, instead of the beacon block root.
+
+### engine_getPayloadBodiesByRangeV1
+
+#### Request
+
+* method: `engine_getPayloadBodiesByRangeV1`
+* params:
+  1. `start`: `QUANTITY`, 64 bits - Starting block number
+  1. `count`: `QUANITTY`, 64 bits - Number of blocks to return
+* timeout: 10s
+
+#### Response
+
+* result: `Array of ExecutionPayloadBodyV1` - Array of [`ExecutionPayloadBodyV1`](#ExecutionPayloadBodyV1) objects.
+* error: code and message set in case an exception happens while processing the method call.
+
+#### Specification
+
+1. Given a `start` and a `count`, the client software **MUST** respond with array of `ExecutionPayloadBodyV1` objects with the corresponding execution block number respecting the order of blocks in the canonical chain, as selected by the latest `engine_forkchoiceUpdated` call.
+
+1. Client software **MUST** support `count` values of at least 32 blocks.
+
+1. Client software **MUST**:
+    * Place `null` in the response array for pruned blocks.
+    * Omit `null` values where the request extends past the current latest known block, i.e. no trailing `nulls` are allowed in this case.
+    * Return array of `null` values if all blocks from requested range are pruned and the range is behind the latest known block.
```

Comment:
this comes out slightly wrong - let's imagine a client that was just checkpointsynced and hasn't downloaded blocks yet - if I make a request that spans the border, it should contain a few nulls, a few blocks and a few "empties"

Reply:
I don't think it is wrong, it is probably written in an unclear way and we might better make the spec statement shorter but provide rich examples to highlight edge cases.

Reply:
Client software **MUST** place `null` in the response array for unavailable blocks which numbers are lower than a number of the latest known block. Client software **MUST NOT** return trailing `null` values if the request extends past the current latest known block. Consider the following response examples:
* `[B1.body, B2.body, ..., Bn.body]` -- entire requested range is filled with block bodies,
* `[null, null, B3.body, ..., Bn.body]` -- first two blocks are unavailable (either pruned or not yet downloaded),
* `[null, null, ..., null]` -- requested range is behind the latest known block and all blocks are unavailable,
* `[B1.body, B2.body, B3.body, B4.body]` -- `B4` is the latest known block,
* `[]` -- requested range is beyond the latest known block,
* `[null, null, B3.body, B4.body]` -- first two blocks are unavailable, `B4` is the latest known block.